File: backend/pkg/database/repository/crawled_content.go
package repository

import (
	"context"
	"encoding/json"
	"fmt"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"github.com/google/uuid"
	"github.com/pgvector/pgvector-go"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// CrawledContentRepository provides data access for crawled content
type CrawledContentRepository struct {
	db *gorm.DB
}

// NewCrawledContentRepository creates a new CrawledContentRepository
func NewCrawledContentRepository(db *gorm.DB) *CrawledContentRepository {
	return &CrawledContentRepository{db: db}
}

// crawledContentModelToType converts a database model to types
func crawledContentModelToType(m *models.CrawledContent) *types.CrawledContent {
	var metaData map[string]interface{}
	if m.MetaData != nil {
		_ = json.Unmarshal(m.MetaData, &metaData)
	}

	var embedding []float32
	if m.Embedding.Slice() != nil {
		embedding = m.Embedding.Slice()
	}

	return &types.CrawledContent{
		ID:           m.ID,
		DataSourceID: m.DataSourceID,
		URL:          m.URL,
		Title:        m.Title,
		Content:      m.Content,
		Summary:      m.Summary,
		Author:       m.Author,
		PublishedAt:  m.PublishedAt,
		Embedding:    embedding,
		MetaData:     metaData,
		CreatedAt:    m.CreatedAt,
	}
}

// crawledContentTypeToModel converts a types type to database model
func crawledContentTypeToModel(c *types.CrawledContent) *models.CrawledContent {
	var metaData datatypes.JSON
	if c.MetaData != nil {
		data, _ := json.Marshal(c.MetaData)
		metaData = datatypes.JSON(data)
	}

	var embedding pgvector.Vector
	if len(c.Embedding) > 0 {
		embedding = pgvector.NewVector(c.Embedding)
	}

	return &models.CrawledContent{
		ID:           c.ID,
		DataSourceID: c.DataSourceID,
		URL:          c.URL,
		Title:        c.Title,
		Content:      c.Content,
		Summary:      c.Summary,
		Author:       c.Author,
		PublishedAt:  c.PublishedAt,
		Embedding:    embedding,
		MetaData:     metaData,
		CreatedAt:    c.CreatedAt,
	}
}

// FindByID retrieves crawled content by its ID
func (r *CrawledContentRepository) FindByID(ctx context.Context, id uuid.UUID) (*types.CrawledContent, error) {
	var model models.CrawledContent
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return crawledContentModelToType(&model), nil
}

// FindByDataSourceID retrieves all crawled content for a data source
func (r *CrawledContentRepository) FindByDataSourceID(ctx context.Context, dataSourceID uuid.UUID, offset, limit int) ([]types.CrawledContent, int64, error) {
	var contentModels []models.CrawledContent
	var total int64

	if err := r.db.WithContext(ctx).Model(&models.CrawledContent{}).Where("data_source_id = ?", dataSourceID).Count(&total).Error; err != nil {
		return nil, 0, err
	}

	if err := r.db.WithContext(ctx).
		Where("data_source_id = ?", dataSourceID).
		Order("created_at DESC").
		Offset(offset).
		Limit(limit).
		Find(&contentModels).Error; err != nil {
		return nil, 0, err
	}

	contents := make([]types.CrawledContent, len(contentModels))
	for i, m := range contentModels {
		contents[i] = *crawledContentModelToType(&m)
	}
	return contents, total, nil
}

// FindByURL checks if content with the given URL exists for a data source
func (r *CrawledContentRepository) FindByURL(ctx context.Context, dataSourceID uuid.UUID, url string) (*types.CrawledContent, error) {
	var model models.CrawledContent
	if err := r.db.WithContext(ctx).Where("data_source_id = ? AND url = ?", dataSourceID, url).First(&model).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return crawledContentModelToType(&model), nil
}

// FindByIDs retrieves multiple crawled content by their IDs
func (r *CrawledContentRepository) FindByIDs(ctx context.Context, ids []uuid.UUID) ([]types.CrawledContent, error) {
	var contentModels []models.CrawledContent
	if err := r.db.WithContext(ctx).Where("id IN ?", ids).Find(&contentModels).Error; err != nil {
		return nil, err
	}

	contents := make([]types.CrawledContent, len(contentModels))
	for i, m := range contentModels {
		contents[i] = *crawledContentModelToType(&m)
	}
	return contents, nil
}

// SearchSimilar performs vector similarity search for crawled content
func (r *CrawledContentRepository) SearchSimilar(ctx context.Context, embedding []float32, limit int) ([]types.CrawledContent, error) {
	var contentModels []models.CrawledContent

	embeddingVector := pgvector.NewVector(embedding)
	query := fmt.Sprintf(
		"SELECT * FROM crawled_content WHERE embedding IS NOT NULL ORDER BY embedding <=> '%s' LIMIT %d",
		embeddingVector.String(),
		limit,
	)

	if err := r.db.WithContext(ctx).Raw(query).Scan(&contentModels).Error; err != nil {
		return nil, err
	}

	contents := make([]types.CrawledContent, len(contentModels))
	for i, m := range contentModels {
		contents[i] = *crawledContentModelToType(&m)
	}
	return contents, nil
}

// SearchSimilarByOrg performs vector similarity search for crawled content within an organization
func (r *CrawledContentRepository) SearchSimilarByOrg(ctx context.Context, orgID uuid.UUID, embedding []float32, limit int) ([]types.CrawledContent, error) {
	var contentModels []models.CrawledContent

	embeddingVector := pgvector.NewVector(embedding)
	query := fmt.Sprintf(
		`SELECT cc.* FROM crawled_content cc 
		 JOIN data_source ds ON cc.data_source_id = ds.id 
		 WHERE ds.organization_id = '%s' AND cc.embedding IS NOT NULL 
		 ORDER BY cc.embedding <=> '%s' LIMIT %d`,
		orgID.String(),
		embeddingVector.String(),
		limit,
	)

	if err := r.db.WithContext(ctx).Raw(query).Scan(&contentModels).Error; err != nil {
		return nil, err
	}

	contents := make([]types.CrawledContent, len(contentModels))
	for i, m := range contentModels {
		contents[i] = *crawledContentModelToType(&m)
	}
	return contents, nil
}

// Save creates new crawled content (upserts based on data_source_id + url)
func (r *CrawledContentRepository) Save(ctx context.Context, content *types.CrawledContent) error {
	model := crawledContentTypeToModel(content)
	if model.ID == uuid.Nil {
		model.ID = uuid.New()
		content.ID = model.ID
	}

	// Upsert: update if exists, create if not
	return r.db.WithContext(ctx).
		Where("data_source_id = ? AND url = ?", model.DataSourceID, model.URL).
		Assign(model).
		FirstOrCreate(model).Error
}

// Update updates existing crawled content
func (r *CrawledContentRepository) Update(ctx context.Context, content *types.CrawledContent) error {
	model := crawledContentTypeToModel(content)
	return r.db.WithContext(ctx).Save(model).Error
}

// Delete removes crawled content by its ID
func (r *CrawledContentRepository) Delete(ctx context.Context, id uuid.UUID) error {
	result := r.db.WithContext(ctx).Delete(&models.CrawledContent{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

// DeleteByDataSourceID removes all crawled content for a data source
func (r *CrawledContentRepository) DeleteByDataSourceID(ctx context.Context, dataSourceID uuid.UUID) error {
	return r.db.WithContext(ctx).Where("data_source_id = ?", dataSourceID).Delete(&models.CrawledContent{}).Error
}

// CountByDataSourceID counts crawled content for a data source
func (r *CrawledContentRepository) CountByDataSourceID(ctx context.Context, dataSourceID uuid.UUID) (int64, error) {
	var count int64
	if err := r.db.WithContext(ctx).Model(&models.CrawledContent{}).Where("data_source_id = ?", dataSourceID).Count(&count).Error; err != nil {
		return 0, err
	}
	return count, nil
}

// FindRecentByOrg retrieves recent crawled content for an organization
func (r *CrawledContentRepository) FindRecentByOrg(ctx context.Context, orgID uuid.UUID, limit int) ([]types.CrawledContent, error) {
	var contentModels []models.CrawledContent

	if err := r.db.WithContext(ctx).
		Joins("JOIN data_source ds ON crawled_content.data_source_id = ds.id").
		Where("ds.organization_id = ?", orgID).
		Order("crawled_content.created_at DESC").
		Limit(limit).
		Find(&contentModels).Error; err != nil {
		return nil, err
	}

	contents := make([]types.CrawledContent, len(contentModels))
	for i, m := range contentModels {
		contents[i] = *crawledContentModelToType(&m)
	}
	return contents, nil
}

File: backend/src/database/repository/crawled_content.rs
use async_trait::async_trait;
use diesel::{
    ExpressionMethods, OptionalExtension, QueryDsl, SelectableHelper, dsl::count_star,
    result::Error as DieselError,
};
use diesel_async::RunQueryDsl;
use pgvector::{Vector, VectorExpressionMethods};
use serde_json::Value;
use uuid::Uuid;

use crate::{
    core::{
        datasource::{CrawledContent, CrawledContentRepository, MetaData},
        insight::InsightContentRepository,
    },
    database::{
        models::crawled_content::{
            CrawledContentChangeset, CrawledContentRow, NewCrawledContentRow,
        },
        pool::PgPool,
    },
    error::AppError,
    schema::{crawled_content, data_source},
};

#[derive(Clone)]
pub struct DieselCrawledContentRepository {
    pool: PgPool,
}

impl DieselCrawledContentRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }

    async fn find_ids(&self, ids: &[Uuid]) -> Result<Vec<CrawledContent>, AppError> {
        let mut connection = self.connection().await?;
        crawled_content::table
            .filter(crawled_content::id.eq_any(ids))
            .select(CrawledContentRow::as_select())
            .load(&mut connection)
            .await
            .map(rows_into_domain)
            .map_err(map_error)
    }

    async fn similar(
        &self,
        organization_id: Option<Uuid>,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        let mut connection = self.connection().await?;
        let vector = Vector::from(embedding.to_vec());
        let mut query = crawled_content::table
            .inner_join(data_source::table)
            .filter(crawled_content::embedding.is_not_null())
            .into_boxed();
        if let Some(id) = organization_id {
            query = query.filter(data_source::organization_id.eq(id));
        }
        if limit >= 0 {
            query = query.limit(limit);
        }
        query
            .order(crawled_content::embedding.cosine_distance(vector))
            .select(CrawledContentRow::as_select())
            .load(&mut connection)
            .await
            .map(rows_into_domain)
            .map_err(map_error)
    }

    async fn recent(
        &self,
        organization_id: Uuid,
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        let mut connection = self.connection().await?;
        let mut query = crawled_content::table
            .inner_join(data_source::table)
            .filter(data_source::organization_id.eq(organization_id))
            .order(crawled_content::created_at.desc())
            .into_boxed();
        if limit >= 0 {
            query = query.limit(limit);
        }
        query
            .select(CrawledContentRow::as_select())
            .load(&mut connection)
            .await
            .map(rows_into_domain)
            .map_err(map_error)
    }
}

#[async_trait]
impl CrawledContentRepository for DieselCrawledContentRepository {
    async fn find_by_data_source_id(
        &self,
        id: Uuid,
        offset: i64,
        limit: i64,
    ) -> Result<(Vec<CrawledContent>, i64), AppError> {
        let mut connection = self.connection().await?;
        let total = crawled_content::table
            .filter(crawled_content::data_source_id.eq(id))
            .select(count_star())
            .first(&mut connection)
            .await
            .map_err(map_error)?;
        let mut query = crawled_content::table
            .filter(crawled_content::data_source_id.eq(id))
            .order(crawled_content::created_at.desc())
            .offset(offset)
            .into_boxed();
        if limit >= 0 {
            query = query.limit(limit);
        }
        let rows = query
            .select(CrawledContentRow::as_select())
            .load(&mut connection)
            .await
            .map_err(map_error)?;
        Ok((rows_into_domain(rows), total))
    }

    async fn find_by_ids(&self, ids: &[Uuid]) -> Result<Vec<CrawledContent>, AppError> {
        self.find_ids(ids).await
    }

    async fn search_similar(
        &self,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.similar(None, embedding, limit).await
    }

    async fn search_similar_by_org(
        &self,
        organization_id: Uuid,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.similar(Some(organization_id), embedding, limit).await
    }

    async fn find_recent_by_org(
        &self,
        organization_id: Uuid,
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.recent(organization_id, limit).await
    }

    async fn find_by_id(&self, id: Uuid) -> Result<CrawledContent, AppError> {
        let mut connection = self.connection().await?;
        crawled_content::table
            .find(id)
            .select(CrawledContentRow::as_select())
            .first(&mut connection)
            .await
            .optional()
            .map_err(map_error)?
            .ok_or(AppError::NotFound)
            .map(Into::into)
    }

    async fn find_by_url(
        &self,
        data_source_id: Uuid,
        url: &str,
    ) -> Result<Option<CrawledContent>, AppError> {
        let mut connection = self.connection().await?;
        crawled_content::table
            .filter(crawled_content::data_source_id.eq(data_source_id))
            .filter(crawled_content::url.eq(url))
            .select(CrawledContentRow::as_select())
            .first(&mut connection)
            .await
            .optional()
            .map(|row| row.map(Into::into))
            .map_err(map_error)
    }

    async fn save(&self, content: &mut CrawledContent) -> Result<(), AppError> {
        if content.id.is_nil() {
            content.id = Uuid::new_v4();
        }
        let now = chrono::Utc::now();
        let row = new_row(content, now);
        let changes = changeset(content);
        let mut connection = self.connection().await?;
        diesel::insert_into(crawled_content::table)
            .values(row)
            .on_conflict((crawled_content::data_source_id, crawled_content::url))
            .do_update()
            .set(changes)
            .returning(crawled_content::id)
            .get_result::<Uuid>(&mut connection)
            .await
            .map(|persisted_id| {
                content.id = persisted_id;
            })
            .map_err(map_error)
    }

    async fn update(&self, content: &CrawledContent) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::update(crawled_content::table.find(content.id))
            .set(changeset(content))
            .execute(&mut connection)
            .await
            .map_err(map_error)?;
        if affected == 0 {
            diesel::insert_into(crawled_content::table)
                .values(new_row(content, chrono::Utc::now()))
                .execute(&mut connection)
                .await
                .map_err(map_error)?;
        }
        Ok(())
    }

    async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        match diesel::delete(crawled_content::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(map_error)?
        {
            0 => Err(AppError::NotFound),
            _ => Ok(()),
        }
    }

    async fn delete_by_data_source_id(&self, id: Uuid) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        diesel::delete(crawled_content::table.filter(crawled_content::data_source_id.eq(id)))
            .execute(&mut connection)
            .await
            .map(|_| ())
            .map_err(map_error)
    }

    async fn count_by_data_source_id(&self, id: Uuid) -> Result<i64, AppError> {
        let mut connection = self.connection().await?;
        crawled_content::table
            .filter(crawled_content::data_source_id.eq(id))
            .select(count_star())
            .first(&mut connection)
            .await
            .map_err(map_error)
    }
}

#[async_trait]
impl InsightContentRepository for DieselCrawledContentRepository {
    async fn find_by_ids(&self, ids: &[Uuid]) -> Result<Vec<CrawledContent>, AppError> {
        self.find_ids(ids).await
    }

    async fn search_similar(
        &self,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.similar(None, embedding, limit).await
    }

    async fn search_similar_by_org(
        &self,
        organization_id: Uuid,
        embedding: &[f32],
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.similar(Some(organization_id), embedding, limit).await
    }

    async fn find_recent_by_org(
        &self,
        organization_id: Uuid,
        limit: i64,
    ) -> Result<Vec<CrawledContent>, AppError> {
        self.recent(organization_id, limit).await
    }
}

impl From<CrawledContentRow> for CrawledContent {
    fn from(row: CrawledContentRow) -> Self {
        Self {
            id: row.id,
            data_source_id: row.data_source_id,
            url: row.url,
            title: row.title,
            content: row.content,
            summary: row.summary,
            author: row.author,
            published_at: row.published_at,
            embedding: row.embedding.map(|value| value.to_vec()),
            meta_data: metadata(row.meta_data),
            created_at: row.created_at,
        }
    }
}

fn new_row(content: &CrawledContent, now: chrono::DateTime<chrono::Utc>) -> NewCrawledContentRow {
    NewCrawledContentRow {
        id: content.id,
        data_source_id: content.data_source_id,
        url: content.url.clone(),
        title: content.title.clone(),
        content: content.content.clone(),
        summary: content.summary.clone(),
        author: content.author.clone(),
        published_at: content.published_at,
        embedding: content.embedding.clone().map(Vector::from),
        meta_data: metadata_value(content.meta_data.as_ref()),
        created_at: content.created_at.unwrap_or(now),
    }
}

fn changeset(content: &CrawledContent) -> CrawledContentChangeset {
    CrawledContentChangeset {
        data_source_id: content.data_source_id,
        url: content.url.clone(),
        title: content.title.clone(),
        content: content.content.clone(),
        summary: content.summary.clone(),
        author: content.author.clone(),
        published_at: content.published_at,
        embedding: content.embedding.clone().map(Vector::from),
        meta_data: content
            .meta_data
            .as_ref()
            .map(|value| metadata_value(Some(value))),
    }
}

fn rows_into_domain(rows: Vec<CrawledContentRow>) -> Vec<CrawledContent> {
    rows.into_iter().map(Into::into).collect()
}

fn metadata(value: Option<Value>) -> Option<MetaData> {
    value.and_then(|value| match value {
        Value::Object(map) => Some(map.into_iter().collect()),
        _ => None,
    })
}

fn metadata_value(value: Option<&MetaData>) -> Value {
    Value::Object(
        value
            .map(|map| map.clone().into_iter().collect())
            .unwrap_or_default(),
    )
}

fn map_error(_: DieselError) -> AppError {
    AppError::Database
}
